<launch>

    <!-- MASTER NODE! -->
    <node name="ros_test_node" pkg="linkalibr" type="ros_test_node" output="screen" clear_params="true" required="true">

        <!-- bag topics -->
        <param name="topic_imu"    type="string" value="/vectornav/IMU" />
        <param name="topic_lidar"  type="string" value="/os_cloud_node/points" />

        <!-- bag parameters -->
        <param name="path_bag"    type="string" value="/home/dji/Downloads/Rosbag/ouster128_IMU.bag" />
        <param name="bag_start"   type="double" value="0.0" />
        <param name="bag_durr"    type="int"    value="-1" />

        <!-- Estimator Parameters -->
        <param name="use_fej" type="bool" value="false" />
        <param name="use_imuavg" type="bool" value="true" />
        <param name="use_rk4int" type="bool" value="true" />
        <param name="calib_lidar_timeoffset" type="bool" value="false" />
        <param name="calib_extrinsics"   type="bool"   value="true" />
        <param name="max_clones" type="int" value="2" />
        <param name="updater_do_chi2_check" type="bool" value="false"/>

        <!-- Filter Init Parameters -->
        <param name="init_window_time" type="double" value="0.5"/> <!-- in second -->
        <param name="init_imu_thresh" type="double" value="0.25"/> <!-- jerk threshold -->

        <!-- Noise Parameters -->
<!--         <param name="accelerometer_noise_density"  type="double"   value="0.00094666" /> -->
<!--         <param name="accelerometer_random_walk"    type="double"   value="0.00278617" /> -->
        <param name="accelerometer_noise_density"  type="double"   value="0.094666" />
        <param name="accelerometer_random_walk"    type="double"   value="0.278617" />
        <param name="gyroscope_noise_density"      type="double"   value="0.00006243" />
        <param name="gyroscope_random_walk"        type="double"   value="0.00000046" />

        <!-- Updater Parameters -->
        <!-- Chi2 Parameters -->
        <param name="updater_chi2_multiplier" type="int" value="1" />
        <!-- Update Noise (sigma) Params -->
        <param name="updater_rotation_noise" type="double" value="0.1" />
        <param name="updater_translation_noise" type="double" value="0.1" />

        <!-- State Initial Noise (sigma) Params -->
        <param name="state_init_x_noise" type="double" value="0.5" />
        <param name="state_init_y_noise" type="double" value="0.5" />
        <param name="state_init_z_noise" type="double" value="0.5" />
        <param name="state_init_rx_noise" type="double" value="0.1" />
        <param name="state_init_ry_noise" type="double" value="0.1" />
        <param name="state_init_rz_noise" type="double" value="0.1" />

        <param name="state_init_timeoffset_noise" type="double" value="0.01" />

        <!-- Gravity -->
        <rosparam param="gravity">[0.0, 0.0, 9.81]</rosparam>

        <!-- initialization of lidar imu time offset -->
        <param name="calib_lidarimu_dt" type="double" value="0.0" />

        <!-- params determining map size -->
        <param name="limit_map_size" type="bool" value="true"/>
        <param name="no_of_scans_for_map" type="int" value="200"/>

        <!-- params for downsampling clouds before mapping -->
        <param name="downsample_for_mapping" type="bool" value="true"/>

        <!-- initial calibration file -->
        <param name="init_calibration_result_filename" type="string" value="$(find linkalibr)/data/I_T_L_init.txt"/>

        <!-- output filenames -->
        <param name="lin_trajectory_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/lin_trajectory"/>
        <param name="lin_bias_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/lin_bias"/>
        <param name="lin_velocity_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/lin_velocity"/>
        <param name="lin_calib_extrinsic_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/lin_calib_extrinsics"/>
        <param name="lin_calib_dt_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/lin_calib_dt"/>

        <param name="lo_trajectory_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/lo_trajectory"/>

        <!-- NDT Resolution -->
        <param name="ndt_resolution" type="double" value="0.25"/>

        <!-- Undistortion flag-->
        <param name="do_undistortion" type="bool" value="true" />

        <!-- Do planar update -->
        <param name="do_planar_update" type="bool" value="false"/>

        <!-- Write an output -->
        <param name="calibration_result_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/I_T_L_final.txt"/>

       <!-- location to store cloud -->
       <param name="raw_scan_folder_name" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/scans/raw"/>
       <param name="deskewed_scan_folder_name" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/scans/deskewed"/>

       <!-- params for batch optimization -->
       <param name="gen_data_for_optimization" type="bool" value="false"/>

       <param name="plane_params_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/plane_params.csv"/>
       <param name="planar_points_raw_scans_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/planar_points_raw_scans.csv"/>
       <param name="planar_points_deskewed_map_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/planar_points_deskewed_map.csv"/>
       <param name="surfel_map_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/surfel_map.csv"/>

       <param name="imu_measurements_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/imu_measurements.csv"/>
       <param name="imu_trajectory_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/imu_trajectory.csv"/>
       <param name="lidar_trajectory_filename" type="string" value="/home/dji/workspace/Texas_LICalib_ws/src/imu_lidar_calibration/linkalibr/data/forOptimization/lidar_trajectory.csv"/>

       <param name="view_surfel_map" type="bool" value="true"/>
    </node>

</launch>
